ledger/smt.go

```diff
@@ -0,0 +1,402 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ledger
+
+// The SMT is derived from https://github.com/aergoio/SMT with modifications
+// to remove unneeded features, and to support retention of old nodes for a fixed time.
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"time"
+
+	"istio.io/pkg/cache"
+)
+
+// TODO when using the SMT, make sure keys and values are same length as Hash
+
+// SMT is a sparse Merkle tree.
+type SMT struct {
+	// Root is the current root of the smt.
+	Root []byte
+	// defaultHashes are the default values of empty trees
+	defaultHashes [][]byte
+	// db holds the cache and related locks
+	db *CacheDB
+	// hash is the hash function used in the trie
+	hash func(data ...[]byte) []byte
+	// TrieHeight is the number if bits in a key
+	TrieHeight int
+	// the minimum length of time old nodes will be retained.
+	retentionDuration time.Duration
+	// lock is for the whole struct
+	lock sync.RWMutex
+	// atomicUpdate, commit all the changes made by intermediate update calls
+	atomicUpdate bool
+}
+
+// this is the closest time.Duration comes to Forever, with a duration of ~145 years
+// we can'tree use int64 max because the duration gets added to Now(), and the ints
+// rollover, causing an immediate expiration (ironic, eh?)
+const forever time.Duration = 1<<(63-1) - 1
+
+// NewSMT creates a new SMT given a keySize, hash function, cache (nil will be defaulted to TTLCache), and retention
+// duration for old nodes.
+func NewSMT(root []byte, hash func(data ...[]byte) []byte, updateCache cache.ExpiringCache, retentionDuration time.Duration) *SMT {
```

Comment:
Do you intend this layer of the API to be used from the outside, or should it be an implementation detail of the ledger? If this should be hidden, then all the symbol names should be lowercased. If they should be exposed, then the name of the package (ledger) doesn't seem appropriate.

Reply:
Good Point.  I will stop exporting this...